src/stores/CatalogOnlineQueryProfileStore.ts

```diff
@@ -0,0 +1,222 @@
+                    display = true;
+                }
+                let controlHeader: ControlHeader = {columnIndex: header.columnIndex, dataIndex: index, display: display, filter: "", columnWidth: null};
+                controlHeaders.set(header.name, controlHeader);
+            }
+        }
+        return controlHeaders;
+    }
+
+    @computed get updateRequestDataSize() {
+        return this.catalogFilterRequest;
+    }
+
+    // do not need infinite scroll for API data
+    @computed get shouldUpdateData(): boolean {
+        return false;
+    }
+
+    @computed get loadOntoImage() {
+        return this.loadingData;
+    }
+
+    @computed get maxRows(): number {
+        return this.numVisibleRows;
+    }
+
+    @action setSortingInfo(columnName: string, sortingType: CARTA.SortingType, columnIndex?: number) {
+        this.sortingInfo = {columnName, sortingType};
+        this.updateSortedIndexMap();
+    }
+
+    @action updateSortedIndexMap() {
+        const dataIndex = this.catalogControlHeader.get(this.sortingInfo.columnName)?.dataIndex;
+        if (dataIndex) {
+            let direction = 0;
+            const sortingType = this.sortingInfo.sortingType;
+            if (sortingType === 0) {
+                direction = 1;
+            } else if (sortingType === 1) {
+                direction = -1;
+            }
+            if (direction === 0 && this.getUserFilters().length === 0) {
+                this.initSortedIndexMap();
+                return;
+            }
+            let catalogColumn = this.catalogData.get(dataIndex);
+            if (this.hasFilter) {
+                this.initSortedIndexMap();
+            }
+            switch (catalogColumn?.dataType) {
+                case CARTA.ColumnType.String:
+                    this.sortedIndexMap.sort((a: number, b: number) => {
+                        const aString = catalogColumn.data[a] as string;
+                        const bString = catalogColumn.data[b] as string;
+                        if (!aString) {
+                            return direction * -1;
+                        }
+
+                        if (!bString) {
+                            return direction * 1;
+                        }
+                        return direction * aString.localeCompare(bString);
+                    });
+                    break;
+                case CARTA.ColumnType.UnsupportedType:
+                    console.log("Data type is not supported");
+                    break;
+                default:
+                    this.sortedIndexMap.sort((a: number, b: number) => {
+                        const aNumber = catalogColumn.data[a] as number;
+                        const bNumber = catalogColumn.data[b] as number;
+                        return direction * (aNumber < bNumber ? -1 : 1);
+                    });
+                    break;
+            }
+        }
+    }
+
+    @action initSortedIndexMap() {
+        this.sortedIndexMap = [];
```

Comment:
shouldn't `sortedIndexMap` be a `Map<>`?

Reply:
the order of this array is represented the sort order. And the value is the index in the original catalog data 😄. Should we rename it to `sortedIndexArray`